Pensel/firmware/applications/pensel_v2.c

```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
-        // let the user know roughly how many workloops are ocurring
-        if (subcount == 100000) {
+        // toggle LED every 10,000 itterations over the workloop
+        if (ledi++ >= 10000) {
             LED_toggle(LED_1);
-            subcount = 0;
-        } else {
-            subcount++;
+            ledi = 0;
+        }
+        // UART_sendint(bDeviceState);
+        if (bDeviceState == CONFIGURED) {
+            CDC_Receive_DATA();
+            /*Check to see if we have data yet */
+            if (Receive_length  != 0) {
+                if (packet_sent == 1)
+                    CDC_Send_DATA ((unsigned char*)Receive_Buffer,Receive_length);
+                Receive_length = 0;
+            }
         }
+    }
 
-        // --- TODO: Check Mag Data ----
+    // initalize the scheduler and add some periodic tasks
+    scheduler_init(&main_schedule);
+    scheduler_add(&main_schedule, 0, heartbeat, &i);
+    scheduler_add(&main_schedule, 0, workloop_flash, &i);
+    scheduler_add(&main_schedule, 0, button_handler, &i);
+    #ifdef WATCHDOG_ENABLE
+        scheduler_add(&main_schedule, 0, watchdog_pet, &i);
+    #endif
 
+    UART_sendString("Initialized\r\n");
 
+    uint32_t sleep_until = 0;
+    while (true) {
+        sleep_until = HAL_GetTick();  // grab start of loop time
+        time_until_next_cb_ms = scheduler_run(&main_schedule, sleep_until);
+        sleep_until += time_until_next_cb_ms;
+
+        if (time_until_next_cb_ms > 0) {
+            UART_sendString("Sleeping for "); UART_sendint(time_until_next_cb_ms); UART_sendString(" ms\r\n");
+            // TODO: sleep instead of dumb delay
+            if (HAL_GetTick() < sleep_until ) {
+                HAL_Delay( sleep_until - HAL_GetTick() );
+            }
+        }
     } /* while (true) */
 } /* main() */
 
-/*! millisecond ISR that increments the global time as well as calls some functions
- *  that need to be periodically serviced. We do a few things:
- *
- *     1. Take care of button and switch debouncing every 10 ms
- *        - If we've changed states in interrupt context and enough time has elapsed (``)
- *          we will change the button / switch state
- *     2. Pet the watchdog every 5 ms. Must be pet within 10 ms!
- *     3. Toggle the heartbeat LED every 1 second
- */
-void HAL_IncTick(void)
-{
-    // sub counter to take care of some tasks every N ms
-    static uint8_t sub_count = 0;
-    static uint16_t second_count = 0;
-
-    // increment the ms timer
-    uwTick++;
-
-    // take care of some things that need to be called periodically every ~10 ms
-    if (sub_count >= 9) {
-        sub_count = 0;
-        button_periodic_handler(uwTick);
-    } else {
-        sub_count++;
-    }
 
-    // every 5 ms, pet the watchdog. Need to pet the watchdog every 10 ms!
-    #ifdef WATCHDOG_ENABLE
-    if (sub_count == 5 && gPetWdg) {
-        wdg_pet();
+/* --- Some common callbacks to be run --- */
+
+ret_t heartbeat(int32_t * new_callback_time_ms) {
```

Comment:
You jump between having the function curly brace on the same line as the declaration and on the following line, just a heads up for consistency